### ISSUE TITLE
Possible rep pen, freq pen, presence pen implementation issues

### ISSUE THREAD

**Issue by kingbri1** [oobabooga/text-generation-webui#4433]
### Describe the bug

The new rep pen implementation which includes presence and frequency penalty in #4376 has resulted in an unusable experience with some parameters set. For example, setting frequency penalty to 0.7 causes responses to become incoherent. This trend becomes worse the further one fills the context with even values such as 0.1 causing gibberish responses.

I have attached comparison tests between Aphrodite engine and oobabooga (using SillyTavern as a frontend), there are definitely some issues here with top-k = 1 tests. Ideally, the responses across regenerations should be the same or slightly differing in wording.

I believe the issue stems from broken penalties. I believe these penalties are applied on normalized logits rather than raw, unnormalized ones. Aphrodite applies these penalties on the unnormalized logits.

The settings are as follows:
Constants:
- Temperature = 0.95
- Top-k = 1

Freq/presence pen:
- Freq pen - 0.7
- Presence pen - 0.7

Rep pen:
- Repetition pen - 1.10

CC @AlpinDale for helping me understand the logic behind this issue (and hopefully he can provide some input on how to fix it).

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Load an AWQ model on ooba, test with top k = 1 (can use my settings if you want). Then load a model on a different backend such as Aphrodite and observe the changes.

The expected results are that the response is the same or has very small variations in wording across regenerations (due to top-k being 1).

### Screenshot

All of these are top-k = 1

Freq pen and pres pen at 0.7
Examples with oobabooga:
![image (5)](https://github.com/oobabooga/text-generation-webui/assets/8082010/b0fc048a-7955-45b0-b3e5-439990b41769)
![image (6)](https://github.com/oobabooga/text-generation-webui/assets/8082010/6f2944d7-80f3-4e87-b973-8d8b11f5abaa)
![image (7)](https://github.com/oobabooga/text-generation-webui/assets/8082010/813b121b-4481-42e5-9ee1-a2649a06e55b)

Examples with Aphrodite:
![image (8)](https://github.com/oobabooga/text-generation-webui/assets/8082010/532aa26e-4139-48f8-9cba-79bd27e7cc1f)

(All 3 tests produced the same response, which they should due to top k = 1).

Rep pen
Examples with oobabooga:

![image (1)](https://github.com/oobabooga/text-generation-webui/assets/8082010/119a7024-f3cc-4509-94cf-efe23f21a877)
![image (2)](https://github.com/oobabooga/text-generation-webui/assets/8082010/db4d2691-44cf-4eeb-9a1a-51938caa48f9)
![image (3)](https://github.com/oobabooga/text-generation-webui/assets/8082010/b82b36ce-efb9-407e-ac7a-285ece758696)

Examples with Aphrodite:
![image (4)](https://github.com/oobabooga/text-generation-webui/assets/8082010/41fb9f7e-60f5-4066-8d70-d6f34edc214b)

(Again, all 3 tests produced the same response, which they should due to top k = 1).

### Logs

```shell
N/A
```

### System Info

```shell
Not really relevant, but Windows 11, RTX 3090ti. Echidna v0.3 AWQ quantized by TheBloke
```

**Comment by oobabooga** [comment]
top_k = 1 concerns sampling, and repetition penalty is a logits processor. The 2 things are independent in the transformers library. repetiton penalty is applied even when do_sample = False, and it can change the top token.

**Comment by bdashore3** [comment]
Top-k is not the issue here. It's used as a testing method for reproducibility of the same response across generations. Even when I have top-k off, the rep pen, freq pen, and presence pen problems still persist. I can attach tests, but they won't really amount to much since the tests also have to persist across subsequent regenerations/swipes.

I want to clarify that the issue is purely rep, freq, and presence pen implementation having possible problems compared to other backends such as Aphrodite engine.

In addition, I believe the rep pen issue is due to fused attention in AutoAWQ. Here's some more tests that I performed with fused attention off and on. When fused attention is off with rep pen at 1.1 and top-k at 1 (for consistency's sake as stated above), the responses behave the same way to the aphrodite tests.

Another thing to note is I renamed this issue since I'm still doing some testing between the two backends.

Without fused_attention (it's the same for all 3 regenerations):
![image (9)](https://github.com/oobabooga/text-generation-webui/assets/8082010/cbed4e9a-7469-42ef-9d53-6ebf2402947a)

With fused_attention:
![image (11)](https://github.com/oobabooga/text-generation-webui/assets/8082010/036f9479-12f6-4966-8d55-2ccd95b45dc5)
![image (12)](https://github.com/oobabooga/text-generation-webui/assets/8082010/56f0e00f-d195-4e54-858d-a373bba94e18)
![image (13)](https://github.com/oobabooga/text-generation-webui/assets/8082010/67f255e8-c1a5-491f-bf98-ec50cb24b68e)

I also get responses such as this with fused_attention enabled:
![image (14)](https://github.com/oobabooga/text-generation-webui/assets/8082010/6274b248-2a13-41b5-8616-2a8358871f19)
